tests/test_optimize_shot_adapative.py

```diff
@@ -0,0 +1,417 @@
+
+        with pytest.raises(ValueError, match=f"The learning rate must be less than {2 / lipschitz}"):
+            opt.step(expval_cost, 0.5)
+
+        # for a single QNode, the lipschitz constant is simply 1
+        opt = qml.ShotAdaptiveOptimizer(min_shots=10, stepsize=100.)
+        with pytest.raises(ValueError, match=f"The learning rate must be less than {2 / 1}"):
+            opt.step(expval_cost.qnodes[0], 0.5)
+
+    def test_unknown_objective_function(self):
+        """Test that an exception is raised if an unknown objective function is passed"""
+        dev = qml.device("default.qubit", wires=1, analytic=False)
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        def cost(x):
+            return np.sin(circuit(x))
+
+        opt = qml.ShotAdaptiveOptimizer(min_shots=10)
+
+        # test expval cost
+        with pytest.raises(ValueError, match="The objective function must either be encoded as a single QNode"):
+            opt.step(cost, 0.5)
+
+        # defining the device attribute allows it to proceed
+        cost.device = circuit.device
+        new_x = opt.step(cost, 0.5)
+
+        assert isinstance(new_x, float)
+
+
+class TestSingleShotGradientIntegration:
+    """Integration tests to ensure that the single shot gradient is correctly computed
+    for a variety of argument types."""
+
+    dev = qml.device("default.qubit", wires=1, analytic=False)
+    H = qml.Hamiltonian([1.0], [qml.PauliZ(0)])
+
+    ansatz = lambda x, **kwargs: qml.RX(x, wires=0)
+    expval_cost = qml.ExpvalCost(ansatz, H, dev)
+
+    @qml.qnode(dev)
+    def qnode(x):
+        qml.RX(x, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+    @pytest.mark.parametrize("cost_fn", [qnode, expval_cost])
+    def test_single_argument_step(self, cost_fn, mocker, monkeypatch):
+        """Test that a simple QNode with a single argument correctly performs an optimization step,
+        and that the single-shot gradients generated have the correct shape"""
+
+        opt = qml.ShotAdaptiveOptimizer(min_shots=10)
+        spy_single_shot_expval = mocker.spy(opt, "_single_shot_expval_gradients")
+        spy_single_shot_qnodes = mocker.spy(opt, "_single_shot_qnode_gradients")
+        spy_grad = mocker.spy(opt, "compute_grad")
+
+        x_init = 0.5
+        new_x = opt.step(cost_fn, x_init)
+
+        assert isinstance(new_x, float)
+        assert new_x != x_init
+
+        spy_grad.assert_called_once()
+
+        if isinstance(cost_fn, qml.ExpvalCost):
+            spy_single_shot_expval.assert_called_once()
+            single_shot_grads = opt._single_shot_expval_gradients(cost_fn, [x_init], {})
+        else:
+            spy_single_shot_qnodes.assert_called_once()
+            single_shot_grads = opt._single_shot_qnode_gradients(cost_fn, [x_init], {})
+
+        # assert single shot gradients are computed correctly
+        assert len(single_shot_grads) == 1

```

Comment:
I wonder why it is nested? Because the qnode returns a scalar output?

Reply:
Yep - the way I coded it, `single_shot_grads` is always a list, even if the function has only one argument